File: android/src/main/java/io/manbang/frontend/thresh/runtime/ThreshEngine.java
/*
 * MIT License
 *
 * Copyright (c) 2020 ManBang Group
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
package io.manbang.frontend.thresh.runtime;

import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.util.Log;

import com.eclipsesource.v8.JavaCallback;
import com.eclipsesource.v8.V8Array;
import com.eclipsesource.v8.V8Object;

import androidx.annotation.Nullable;

import io.manbang.frontend.thresh.runtime.jscore.bundle.BundleCallback;
import io.manbang.frontend.thresh.runtime.jscore.bundle.JSBundleLoader;
import io.manbang.frontend.thresh.runtime.jscore.runtime.JSCallback;
import io.manbang.frontend.thresh.runtime.jscore.util.V8Util;
import io.manbang.frontend.thresh.channel.BridgeCallback;
import io.manbang.frontend.thresh.channel.MethodConstants;
import io.manbang.frontend.thresh.channel.nativemodule.NativeModule;
import io.manbang.frontend.thresh.channel.MethodChannelModule;
import io.manbang.frontend.thresh.manager.ContextIdManager;
import io.manbang.frontend.thresh.util.ThreshLogger;

import java.util.Locale;
import java.util.Map;

import io.manbang.frontend.thresh.view.LifecycleListener;

import static io.manbang.frontend.thresh.channel.MethodConstants.CALL_TIMER_OPERATOR;

/**
 * thresh engine
 */
public class ThreshEngine implements LifecycleListener, EngineService {

    /**
     * bundle options
     */
    private ThreshEngineOptions engineOptions;
    private String contextId;
    /**
     * channel
     */
    private MethodChannelModule threshChannel;
    /**
     * native module
     */
    private NativeModule nativeModule;
    /**
     * bundle loader
     */
    public JSBundleLoader bundleLoader;
    /**
     * handler
     */
    private Handler mMainHandler;
    /**
     * JS module
     */
    public JSModule jsModule;

    /**
     * thresh engine status
     *
     * <p> IF isReady true : thresh engine ok ,else fail</p>
     */

    public ThreshEngine(ThreshEngineOptions engineOptions, String contextId) {
        this.engineOptions = engineOptions;
        this.contextId = contextId;
        assertBundleParams();
        jsModule = JSManager.getInstance().getJSModule(engineOptions.moduleName);
        if (jsModuleInvalid()) {
            jsModule.destroy();
            jsModule = null;
        }
        if (jsModule == null) {
            jsModule = new JSModule(engineOptions.moduleName, engineOptions.moduleVersion, engineOptions.bundleOptions);
            JSManager.getInstance().registerJSModule(jsModule);
        }
        this.mMainHandler = new Handler(Looper.getMainLooper());
        registerMethodCall();
    }

    //JSModule失效（存在但不可用）
    private boolean jsModuleInvalid() {
        if (jsModule == null) {
            return false;
        }
        if (!engineOptions.moduleVersion.equals(jsModule.getModuleVersion())) {//版本号不一致
            return true;
        }

        if (!jsModule.sameDebugService(engineOptions.bundleOptions)) {//服务地址不一致
            return true;
        }

        return false;
    }

    /**
     * bind method channel for thresh engine
     */
    public void bindThreshMethodCall(MethodChannelModule channel) {
        this.threshChannel = channel;
    }

    /**
     * bind native bridge for thresh engine
     */
    public void bindNativeModule(NativeModule nativeModule) {
        this.nativeModule = nativeModule;
    }

    /**
     * bind bundle loader for thresh engine
     */
    public void bindBundleLoader(JSBundleLoader bundleLoader) {
        this.bundleLoader = bundleLoader;
    }

    /**
     * register flutter <->  native <-> JS  call
     **/
    private void registerMethodCall() {

        jsModule.registerJavaMethodCallBack(jsCallFlutterCallBack, engineOptions.jsCallFlutterMethod);

        jsModule.registerJavaMethodCallBack(jsCallNativeCallBack, engineOptions.jsCallNativeMethod);

        jsModule.registerJavaMethodCallBack(jsCallTimerCallBack, engineOptions.callNativeTimer);

        jsModule.registerLifecycleEvent(engineOptions.callJSLifecycleMethod, engineOptions.callJSMethod);
    }

    private JavaCallback jsCallFlutterCallBack = new JavaCallback() {
        @Override
        public Object invoke(V8Object v8Object, V8Array v8Array) {
            performCallFlutter(V8Util.toMap(v8Array));
            return null;
        }
    };

    private JavaCallback jsCallNativeCallBack = new JavaCallback() {
        @Override
        public Object invoke(V8Object v8Object, V8Array v8Array) {
            performCallNative(V8Util.toMap(v8Array));
            return null;
        }
    };

    private JavaCallback jsCallTimerCallBack = new JavaCallback() {
        @Override
        public Object invoke(V8Object v8Object, V8Array v8Array) {
            final Map args = V8Util.toMap(v8Array);
            args.put(MethodConstants.CALL_METHOD, CALL_TIMER_OPERATOR);
            performCallNative(args);
            return null;
        }
    };


    /**
     * An method for js call flutter
     * <LI>  {@link ThreshEngineOptions#jsCallFlutterMethod} </LI>
     */
    public void performCallFlutter(final Map args) {
        if (isNotTargetContextId(args)) {
            return;
        }
        mMainHandler.post(new Runnable() {
            @Override
            public void run() {
                if (threshChannel != null) {
                    threshChannel.invokeMethod(engineOptions.jsCallFlutterMethod, args);
                }
            }
        });
    }

    private boolean isNotTargetContextId(Map args) {
        Object methodContextId = args.get("contextId");
        if (methodContextId == null) {
            return false;
        }
        return !this.contextId.equals(methodContextId);
    }

    /**
     * An method for js call native
     * <LI>  {@link ThreshEngineOptions#jsCallNativeMethod} </LI>
     */
    private void performCallNative(Map args) {
        if (isNotTargetContextId(args)) {
            return;
        }
        if (args != null) {
            nativeModule.dispatchJSCallNativeEvent(ThreshEngine.this, args.get(MethodConstants.CALL_METHOD) + "", args);
        }
    }

    /**
     * {@link LifecycleListener#onPause()}
     */
    @Override
    public void onPause() {
        if (jsModule != null) {
            jsModule.onPause(contextId);
        }
    }

    /**
     * {@link LifecycleListener#onResume()}
     */
    @Override
    public void onResume() {
        if (jsModule != null) {
            jsModule.onResume(contextId);
        }
    }

    /**
     * {@link LifecycleListener#onDestroy()}
     */
    @Override
    public void onDestroy() {
        ContextIdManager.INSTANCE.clearByContextId(contextId);
        if (jsModule != null) {
            jsModule.onDestroy(contextId);
            jsModule.unregisterJavaMethodCallBack(jsCallFlutterCallBack, engineOptions.jsCallFlutterMethod);
            jsModule.unregisterJavaMethodCallBack(jsCallNativeCallBack, engineOptions.jsCallNativeMethod);
            jsModule.unregisterJavaMethodCallBack(jsCallTimerCallBack, engineOptions.callNativeTimer);
        }
        // unregister channel
        if (threshChannel != null) {
            threshChannel.unRegister();
            threshChannel = null;
        }
        bundleLoader = null;
        nativeModule = null;
    }

    /**
     * {@link LifecycleListener#onBackPressed()}
     */
    @Override
    public void onBackPressed() {
        if (jsModule != null) {
            jsModule.onBackPressed(contextId);
        }
    }

    public ThreshEngineOptions getThreshOptions() {
        return engineOptions;
    }

    public void setReady(boolean isReady) {
        ThreshLogger.v(String.format(Locale.US, "[--- ThreshEngine READY %s --- ]", isReady ? "Success" : "Failure"));
    }


    public void assertBundleParams() {
        if (engineOptions == null
                || engineOptions.bundleOptions == null
                || TextUtils.isEmpty(engineOptions.moduleName)) {
            ThreshLogger.e("Must be bundle options value exception");
            throw new ThreshException("Must be bundle options value.");
        }
    }

    @Override
    public void loadScript() {
        loadScript(null);
    }

    @Override
    public void forceLoadScript() {
        loadScriptImpl(null);
    }

    @Override
    public void loadScript(final JSCallback callback) {
        if (jsModule.isLoaded()) {
            return;
        }
        loadScriptImpl(callback);
    }

    private void loadScriptImpl(final JSCallback callback) {
        if (bundleLoader == null) {
            if (callback != null) {
                callback.error(-1, "BundleLoader cannot null", "");
            }
            return;
        }
        bundleLoader.loadScript(null, engineOptions.bundleOptions, new BundleCallback() {
            @Override
            public void onResult(int code, String reason, String data) {
                if (code == 0 && !TextUtils.isEmpty(data)) {
                    if (callback != null) {
                        jsModule.executeScript(contextId, data, callback);
                    } else {
                        jsModule.executeScript(contextId, data, new JSCallback() {
                            @Override
                            public void success(@Nullable Object o) {
                                setReady(true);
                            }

                            @Override
                            public void error(int i, @Nullable String s, @Nullable Object o) {
                                setReady(false);
                            }
                        });
                    }
                } else {
                    ThreshLogger.e("READ BUNDLE ERROR:" + reason);
                }
            }
        });
    }


    @Override
    public void loadBundleScript(final BundleCallback callback) {
        if (jsModule.isLoaded()) {
            return;
        }
        if (bundleLoader != null) {
            bundleLoader.loadScript(null, engineOptions.bundleOptions, callback);
        } else {
            if (callback != null) {
                callback.onResult(-1, "BundleLoader cannot null", "");
            }
        }
    }

    @Override
    public void executeScript(final String script, final JSCallback callback) {
        if (jsModule.isLoaded()) {
            return;
        }
        if (callback != null) {
            jsModule.executeScript(contextId, script, callback);
        } else {
            jsModule.executeScript(contextId, script, new JSCallback() {
                @Override
                public void success(@Nullable Object o) {
                    setReady(true);
                }

                @Override
                public void error(int i, @Nullable String s, @Nullable Object o) {
                    setReady(false);
                }
            });
        }
    }

    @Override
    public void execMessage(String method, Object params) {
        if (jsModule != null) {
            jsModule.execMessage(contextId, method, params);
        }
    }

    @Override
    public void executeJSFunction(String funcName, JSCallback callback, Map params) {
        if (jsModule != null) {
            jsModule.executeJSFunction(contextId, funcName, callback, params);
        }
    }

    @Override
    public void invokeNativeModule(String module, String method, Map params, BridgeCallback callback) {
        if (nativeModule != null) {
            nativeModule.invokeNativeModule(module, method, params, callback);
        }
    }
}

File: ios/Classes/runtime/ThreshEngine.m
/**
* MIT License
*
* Copyright (c) 2020 ManBang Group
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*
*/

#import "ThreshEngine.h"
#import "ThreshJSChannelManager.h"
#import "ThreshFlutterPluginChannel.h"
#import "ThreshPlatformViewFactory.h"
#import "ThreshJSChannelManager.h"
#import "ThreshAppDelegate.h"
#import "ThreshJSLoader.h"

static NSMapTable<NSString *, NSHashTable *> *_loadCallbacks; // 装载js回调
static NSMutableDictionary *_viewContextDic;
static NSUInteger contextId = 0;

typedef void(^Callback)(BOOL success);

@interface ThreshEngine () <ThreshJSChannelDelegate, ThreshPluginDelegate>

@property (nonatomic, copy) NSString *pageName;
@property (nonatomic, copy) NSString *moduleName;

@property (nonatomic, weak) id<ThreshProtocol> config;

@property (nonatomic, strong) FlutterEngine *flutterEngine;

@property (nonatomic, strong) ThreshJSCoreChannel *jsCoreChannel;

@property (nonatomic, strong) ThreshPlatformUniqueChannelKeyGenerator *threshGenerator;
@property (nonatomic, strong) ThreshFlutterPluginChannel *flutterThreshChannel;

@property (nonatomic, assign) NSUInteger rootId;

@property (nonatomic, assign) BOOL released;

@end

@implementation ThreshEngine

+ (void)load {
    _viewContextDic = @{}.mutableCopy;
    _loadCallbacks = [[NSMapTable alloc] initWithKeyOptions:NSPointerFunctionsStrongMemory valueOptions:NSPointerFunctionsStrongMemory capacity:1];
}

#pragma mark - init

- (instancetype)initWithConfig:(id<ThreshProtocol>)config flutterEngine:(FlutterEngine *)engine rootId:(NSUInteger)rootId {
    
    if (self = [super init]) {
        self.config = config;
        self.rootId = rootId;
        self.pageName = safeRespondsForProtocol(config, @selector(router), @selector(pageName), nil) ? config.router.pageName : @"errorPage";
        self.moduleName = safeRespondsForProtocol(config, @selector(router), @selector(moduleName), nil) ? config.router.moduleName : @"error";
        
        if (safeRespondsForProtocol(config, @selector(router), @selector(entrypoint), nil)) {
            [engine runWithEntrypoint:config.router.entrypoint];
        }
        
        self.flutterEngine = engine;
        
        ThreshInfo(@"ThreshEngine init, pageName = %@, moduleName = %@", self.pageName, self.moduleName);
        if (engine) {
            [self _registerChannelsWithEngine:engine];
            [self _activateJS];
        } else {
            ThreshWarn(@"ThreshEngine init need Flutter Engine");
        }
    }
    return self;
}

- (void)_registerChannelsWithEngine:(FlutterEngine *)engine {
    
    Class class = NSClassFromString(@"GeneratedPluginRegistrant");
    if (class && [class respondsToSelector:NSSelectorFromString(@"registerWithRegistry:")]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [class performSelector:NSSelectorFromString(@"registerWithRegistry:") withObject:self.flutterEngine];
#pragma clang diagnostic pop
    }
    
    self.jsCoreChannel = [[ThreshJSChannelManager sharedMgr] channelWithModule:self.moduleName];
    self.jsCoreChannel.delegate = self;
    __weak typeof(self) weakSelf = self;
    self.jsCoreChannel.exceptionHandler = ^(NSString * _Nonnull exception) {
        [weakSelf exceptionType:ThreshExceptionJS msg:exception];
    };
    
    [self _registerFlutterPlugin:[engine registrarForPlugin:NSStringFromClass([self class])]];
}

- (void)_activateJS {
    
    // 改变 contextId
    _contextId = ++contextId;
    // 将 rootId 同 contextId 绑定
    _viewContextDic[@(self.rootId)] = @(contextId);
    // 将 jscore 回调同 contextId 绑定
    [self.jsCoreChannel addDelegate:self contextId:contextId];
    // 加载bundle包
    if (![[ThreshJSChannelManager sharedMgr] loadedWithModule:self.moduleName]) {
        ThreshInfo(@"Loader: %@ haven't loaded, begin load", self.moduleName);
        __weak typeof(self) weakSelf = self;
        [self _startLoad:^(BOOL success) {
            if (success) {
                [[ThreshJSChannelManager sharedMgr] haveLoadedWithModule:weakSelf.moduleName];
                ThreshInfo(@"Loader: %@ load success", weakSelf.moduleName);
            } else {
                ThreshInfo(@"Loader: %@ load failed", weakSelf.moduleName);
            }
        }];
    } else {
        ThreshInfo(@"Loader: %@ have loaded", self.moduleName);
    }
}

#pragma mark - load js

- (void)_startLoad:(Callback)callback {
    
    // 记录回调
    if ([[_loadCallbacks keyEnumerator].allObjects containsObject:self.moduleName]) {
        ThreshInfo(@"Loader: %@ loading", self.moduleName);
        NSHashTable *t = [_loadCallbacks objectForKey:self.moduleName];
        [t addObject:callback];
    } else {
        ThreshInfo(@"Loader: %@ first load", self.moduleName);
        NSHashTable *t = [[NSHashTable alloc] initWithOptions:NSPointerFunctionsStrongMemory capacity:1];
        [t addObject:callback];
        [_loadCallbacks setObject:t forKey:self.moduleName];
        
        // 开始加载
        __weak typeof(self) weakSelf = self;
        [self _loadAndEvalJS:^(BOOL success) {
            if (success) {
                NSHashTable *t = [_loadCallbacks objectForKey:weakSelf.moduleName];
                [[t objectEnumerator].allObjects enumerateObjectsUsingBlock:^(Callback  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
                    if (obj) {
                        obj(YES);
                    }
                }];
                [_loadCallbacks removeObjectForKey:weakSelf.moduleName];
            }
        }];
    }
}

- (void)_loadAndEvalJS:(void (^)(BOOL success))complete {
    
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        
        __weak typeof(self) weakSelf = self;
        void (^callback)(NSData *, NSError *) = ^(NSData *response, NSError *err) {
            
            __strong typeof(weakSelf) strongSelf = weakSelf;
            [strongSelf exportLifeCycle:ThreshAfterLoadBundle ext:@{}];
            NSString *dataJSString = [[NSString alloc] initWithData:response encoding:NSUTF8StringEncoding];
            if(dataJSString.length > 0) {
                [strongSelf exportLifeCycle:ThreshBeforeEvalJS ext:@{}];
                [strongSelf.jsCoreChannel evaluateJS:dataJSString callback:^{
                    [strongSelf exportLifeCycle:ThreshAfterEvalJS ext:@{}];
                    complete(YES);
                }];
            } else {
                complete(NO);
                ThreshError(@"load js app error, js string length 0");
            }
        };
        
        [self exportLifeCycle:ThreshBeforeLoadBundle ext:@{}];
        if (safeRespondsForProtocol(self.config, @selector(loader), @selector(loadType), nil) &&
            self.config.loader.loadType == LoadWithURL) {
            [[ThreshJSLoader sharedLoader] loadJSWithAddress:(safeRespondsForProtocol(self.config, @selector(loader), @selector(serverAddress), nil) ? self.config.loader.serverAddress : @"") response:callback];
        } else {
            if (safeRespondsForProtocol(self.config, @selector(loader), @selector(getJSDataWithModuleName:callback:), nil)) {
                [self.config.loader getJSDataWithModuleName:self.moduleName callback:^(NSData *data) {
                    if(callback && data) {
                        callback(data, nil);
                    } else {
                        callback(nil, nil);
                    }
                }];
            } else {
                ThreshWarn(@"need implementation for 'getJSDataWithModuleName:callback:' (at ThreshDataLoader.h)");
            }
        }
    });
}

#pragma mark - flutter plugin
// ThreshPluginDelegate
- (void)threshFlutterPluginCall:(NSString*)method arguments:(id)arguments complete:(ThreshCompleteBlock)callback {
    if ([kChannelFlutterCallJS isEqualToString:method]) {
        ThreshInfo(@"Flutter call MethodChannel_Flutter_call_JS: %@\narguments: %@\n",method, arguments);
        [self _callJS:method arguments:arguments callback:^(NSDictionary *response) {
            if (callback) {
                callback(response);
            }
        }];
    } else if([kChannelFlutterCallNative isEqualToString:method]) {
        [self _callNative:arguments callback:^(NSDictionary *response) {
            ThreshInfo(@"Flutter call MethodChannel_Flutter.method:%@\narguments:%@\nNative response: %@ ", method, arguments, response);
            if (callback) {
                callback(response);
            }
        }];
    } else if ([kFlutterBridgeCallNative isEqualToString:method]) {
        
        if (safeRespondsForProtocol(self.config, @selector(nativeBridge), @selector(performBridge:callBack:), nil)) {
            [self.config.nativeBridge performBridge:arguments callBack:^(NSDictionary *response) {
                if (callback) {
                    callback(response);
                }
            }];
        }
    } else {
        ThreshError(@"flutter call, method channel unrecognized : %@\n arguments:%@", method,arguments);
    }
}

- (void)_registerFlutterPlugin:(NSObject<FlutterPluginRegistrar> *)registrar {
    
    _threshGenerator = [[ThreshPlatformUniqueChannelKeyGenerator alloc] initWithPageName:self.pageName
                                                                       moduleName:self.moduleName
                                                                      channelName:kThreshChannelName];
    self.flutterThreshChannel = [ThreshFlutterPluginChannel pluginWithModel:_threshGenerator];
    self.flutterThreshChannel.delegate = self;
    [ThreshFlutterPluginChannel registerWithRegistrar:registrar];
    
    // 注册自定义nativeView
    if (safeRespondsForProtocol(self.config, @selector(nativePlatform), @selector(platformViewConfig), nil)) {
        NSDictionary *platformViewConfig = self.config.nativePlatform.platformViewConfig;
        NSArray *keys = [platformViewConfig allKeys];
        [keys enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            if ([obj isKindOfClass:[NSString class]]) {
                NSString *viewId = (NSString *)obj;
                NSString *objClassName = platformViewConfig[viewId];
                [registrar registerViewFactory:[[ThreshPlatformViewFactory alloc] initWithRegistrar:registrar objClassName:objClassName] withId:viewId];
            }
        }];
    }
}

- (void)_callFlutter:(NSString *)methodName arguments:(id _Nullable)arguments callback:(ThreshCompleteBlock)completeBlock {
    [self.flutterThreshChannel.channel invokeMethod:methodName arguments:arguments result:^(id  _Nullable result) {
        [self _handleFlutterResult: result];
    }];
}

- (void)_handleFlutterResult:(id  _Nullable) result {
    if([result isKindOfClass:[FlutterError class]]) {
        FlutterError *flutterErr = (FlutterError *)result;
        if(flutterErr) {
            ThreshError(@"Flutter error code: %@, message: %@", flutterErr.code, flutterErr.message);
        }
    } else if([result isKindOfClass:[NSDictionary class]]) {
        NSDictionary *resultDic = (NSDictionary *)result;
        if(resultDic) {
            ThreshInfo(@"flutter invoke result: %@", resultDic);
        }
    } else {
        if (result) {
            ThreshInfo(@"result class : %@", NSStringFromClass([result class]));
        }
    }
}

- (void)_releaseFlutterChannel {
    if(self.flutterThreshChannel) {
        [self.flutterThreshChannel releasePlugin];
        self.flutterThreshChannel = nil;
     }
}

#pragma mark - js channel
// ThreshJSChannelDelegate
- (void)jsChannelCallMethod:(NSString *)method arguments:(id)arguments complete:(ThreshCompleteBlock)callback {
    
    if ([kChannelJSCallFlutter isEqualToString:method]) {
        ThreshInfo(@"methodChannel_js_call_flutter method: %@, params :%@", method, arguments);
        [self _callFlutter:method arguments:arguments callback:^(NSDictionary *response) {}];
    } else if([kChannelJSCallNative isEqualToString:method]) {
        ThreshInfo(@"methodChannel_js_call_native method: %@, params :%@", method, arguments);
        [self _callNative:arguments callback:^(NSDictionary *response) {
            if(callback && response) {
                callback(response);
            }
        }];
    } else {
        ThreshError(@"js channel call native, method channel unrecognized : %@, arguments:%@", method,arguments);
    }
}

- (void)_callJS:(NSString *)methodName arguments:(id _Nullable)arguments callback:(ThreshCompleteBlock)completeBlock {
    [self.jsCoreChannel invokeMothod:methodName args:arguments complete:^(NSDictionary *resultDic) {
        if(resultDic) {
            ThreshInfo(@"flutter invoke methodName: %@ \narguments:%@ \nresult: %@", methodName, arguments, resultDic);
            if (completeBlock) {
                completeBlock(resultDic);
            }
        }
    }];
}

- (void)_releaseJSChannel {
    
    if (self.jsCoreChannel) {
        [self _destroyJSEnv];
        [_viewContextDic removeObjectForKey:@(self.rootId)];
    }
}

// 通知js侧销毁对应的实例
- (void)_destroyJSEnv {
    dispatch_semaphore_t s = dispatch_semaphore_create(0);
    NSMutableDictionary *params = @{@"method": @"onDestroyed"}.mutableCopy;
    if (_viewContextDic[@(self.rootId)]) {
        [params setObject:[NSString stringWithFormat:@"%@", _viewContextDic[@(self.rootId)]] forKey:@"contextId"];
    }
    [self.jsCoreChannel invokeMothod:kChannelNativeCallJS args:params complete:^(NSDictionary *response) {}];
    dispatch_semaphore_wait(s, dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.03 * NSEC_PER_SEC)));
}

#pragma mark - native bridge

- (void)_callNative:(id)arguments callback:(ThreshCompleteBlock)callback {
    
    NSDictionary *dic = (NSDictionary *)arguments;
    if(dic) {
        
        NSString *method = [dic objectForKey:kChannelMethodName];

        if([kMethodReloadJSBundle isEqualToString:method]) {
            //触发js reload
            [self reloadJS];
        } else if([kMethodBridgeRequest isEqualToString:method]){
            //js调用bridge
            NSDictionary *params = [dic objectForKey:kChannelParams];
            ThreshInfo(@"bridgeRequest params :%@", params);
            if(params) {
                NSDictionary *request = params[@"request"];
                if ([request isKindOfClass:[NSDictionary class]] && [request[@"method"] isEqualToString:kMethodCloseWindow]) {
                    [self releaseEngine];
                }
                if (safeRespondsForProtocol(self.config, @selector(nativeBridge), @selector(performBridge:callBack:), nil)) {
                    [self.config.nativeBridge performBridge:request callBack:^(NSDictionary *response) {
                        if(callback && response) {
                            callback(response);
                        }
                    }];
                }
            }
            else {
                ThreshError(@"params is nil");
            }
        } else if([kMethodPageDidLoad isEqualToString:method]
                  ||[kMethodPageDidShow isEqualToString:method]) {
            NSDictionary *params = [arguments isKindOfClass:[NSDictionary class]] ? arguments[kChannelParams] : nil;
            [self exportLifeCycle:ThreshPageDidShow ext:params];
            if ([kMethodPageDidShow isEqualToString:method]) {
                // 如果Thresh异常，用户触发刷新还会走pageDidShow方法，导致脏数据，置为nil，则不会产生新的时间埋点
            }
        } else if ([method isEqualToString:kFlutterBridgeCallNative]) {
            if (safeRespondsForProtocol(self.config, @selector(nativeBridge), @selector(performBridge:callBack:), nil)) {
                [self.config.nativeBridge performBridge:arguments[@"params"] callBack:^(NSDictionary *response) {
                    if(callback && response) {
                        callback(response);
                    }
                }];
            }
        } else if ([method isEqualToString:kInvokePlatformViewMethod]) {
            if (safeRespondsForProtocol(self.config, @selector(nativePlatform), @selector(invokeInstanceMethodWithArgs:), nil)) {
                [self.config.nativePlatform invokeInstanceMethodWithArgs:arguments[@"params"]];
            }
        } else {
            ThreshError(@"unknown method !!!! %@", method);
        }
    } else {
        ThreshError(@"call native, trans to dic error, arguments:%@", arguments);
    }
}

#pragma mark - public method

- (void)viewWillAppear {
    if (self.jsCoreChannel && [self.jsCoreChannel envReady]) {
        [self sendJSEvent:@"pageOnShow" complete:^(NSDictionary *response) {}];
    }
}

- (void)viewWillDisappear {
    if (self.jsCoreChannel && [self.jsCoreChannel envReady]) {
        [self sendJSEvent:@"pageOnHide" complete:^(NSDictionary *response) {}];
    }
}

- (void)reloadJS {
    if (self.jsCoreChannel && [self.jsCoreChannel envReady]) {
        // 清空之前的context 并重新创建
        [self.jsCoreChannel reloadJS];
        // 加载
        [self _startLoad:^(BOOL success) {}];
    }
}

- (void)reloadPage {
    if (self.jsCoreChannel && [self.jsCoreChannel envReady]) {
        // pageOnShow时，JS侧会进行页面数据刷新，起到刷新页面的功能
        [self sendJSEvent:@"pageOnShow" complete:^(NSDictionary *response) {}];
    }
}

- (void)sendJSEvent:(id)eventStr complete:(ThreshCompleteBlock)complete {
    
    NSString *contextId = @"";
    if (_viewContextDic[@(self.rootId)]) {
        contextId = [NSString stringWithFormat:@"%@", _viewContextDic[@(self.rootId)]];
        [self.jsCoreChannel invokeMothod:kChannelFireJSEvent complete:complete withArguments:eventStr, contextId, nil];
    }
}

- (void)sendJSEvent:(id)eventStr args:(NSDictionary *)args complete:(ThreshCompleteBlock)complete {
    NSString *contextId = @"";
    if (_viewContextDic[@(self.rootId)]) {
        contextId = [NSString stringWithFormat:@"%@", _viewContextDic[@(self.rootId)]];
        [self.jsCoreChannel invokeMothod:kChannelFireJSEvent complete:complete withArguments:eventStr, contextId, args, nil];
    }
}

- (void)invokeJSMethod:(NSString *)methodStr args:(id)args complete:(ThreshCompleteBlock)complete {
    [self.jsCoreChannel invokeMothod:methodStr args:args complete:complete];
}

- (void)releaseEngine {
    // 此处只会释放一次，需要添加标识
    if (!self.released) {
        self.released = YES;
        [self _releaseJSChannel];
        [self _releaseFlutterChannel];
    }
}

#pragma mark - private method

- (void)exportLifeCycle:(ThreshLifeCycle)lifeCycle ext:(NSDictionary *)ext {
    if (safeRespondsForProtocol(self.config, @selector(exporter), @selector(lifeCycle:ext:), nil)) {
        [self.config.exporter lifeCycle:lifeCycle ext:ext];
    }
}

- (void)exceptionType:(ThreshExceptionType)exceptionType msg:(NSString *)msg {
    if (safeRespondsForProtocol(self.config, @selector(exporter), @selector(exceptionType:msg:), nil)) {
        [self.config.exporter exceptionType:exceptionType msg:msg];
    }
}

- (void)dealloc {
    ThreshInfo(@"ThreshEngine dealloc, pageName = %@, moduleName = %@", self.pageName, self.moduleName);
}

@end
